src/riot-rs-threads/src/lib.rs

```diff
@@ -227,40 +299,165 @@ impl Threads {
+        #[cfg(not(feature = "multicore"))]
         match self.current().map(|t| t.prio) {
             Some(curr_prio) if curr_prio < prio => schedule(),
             _ => {}
         }
+        #[cfg(feature = "multicore")]
+        match self.lowest_running_prio(_thread_id) {
+            (core, Some(lowest_prio)) if lowest_prio < prio => schedule_on_core(core),
+            _ => {}
+        }
+    }
+
+    /// Returns `Some` if the thread is currently running on a core.
+    ///
+    /// On multicore, the core-id is returned as usize, on single-core
+    /// the usize is always 0.
+    fn is_running(&self, thread_id: ThreadId) -> Option<usize> {
+        #[cfg(not(feature = "multicore"))]
+        {
+            self.current_pid()
+                .filter(|pid| *pid == thread_id)
+                .map(|_| 0)
+        }
+
+        #[cfg(feature = "multicore")]
+        {
+            self.current_threads
+                .iter()
+                .position(|pid| *pid == Some(thread_id))
+        }
+    }
+
+    /// Adds the thread that is running on the current core to the
+    /// runqueue if it has state [`ThreadState::Running`].
+    #[cfg(feature = "multicore")]
+    #[allow(dead_code, reason = "used in scheduler implementation")]
+    fn add_current_thread_to_rq(&mut self) {
+        let (pid, prio) = match self.current() {
+            Some(&mut Thread {
+                pid,
+                prio,
+                state: ThreadState::Running,
+                ..
+            }) => (pid, prio),
+            _ => return,
+        };
+        self.runqueue.add(pid, prio);
+    }
+
+    /// Returns the next thread from the runqueue.
+    ///
+    /// On single-core, the thread remains in the runqueue, so subsequent calls
+    /// will return the same thread.
+    ///
+    /// On multi-core, the thread is removed so that subsequent calls will each
+    /// return a different thread. This prevents that a thread is picked multiple
+    /// times by the scheduler when it is invoked on different cores.
+    #[allow(dead_code, reason = "used in scheduler implementation")]
+    fn get_next_pid(&mut self) -> Option<ThreadId> {
+        // On single-core, only read the head of the runqueue.
+        #[cfg(not(feature = "multicore"))]
+        {
+            self.runqueue.get_next()
+        }
+
+        // On multi-core, the head is popped of the runqueue.
+        #[cfg(all(feature = "multicore", not(feature = "core-affinity")))]
+        {
+            self.runqueue.pop_next()
+        }
+
+        // On multicore with core-affinities, get next thread with matching affinity.
+        #[cfg(all(feature = "multicore", feature = "core-affinity"))]
+        {
+            let next = self
+                .runqueue
+                .get_next_filter(|&t| self.is_affine_to_curr_core(t))?;
+            // Delete thread from runqueue to match the `pop_next`.
+            self.runqueue.del(next);
```

Comment:
Not sure I understand what you mean. Where is it iterating twice?

Reply:
runqueue is a linked list, so `get_next_filter()` has to iterate the list in order to find `next`. `del(next)` now has to iterate again to unlink `next`. If I understand correctly.
Textbook deletion would save one of the iterations by finding `next` and unlinking directly. 
(if at all, this is a minor optimisation. If my understanding is correct, maybe add "// TODO: this would benefit from `del_one_with_filter(...)`" or similar.)

Reply:
The kind of iterations we're doing is a bit different between the two, but agreed that this can be optimized.
Will add the TODO.